System.IO.FileSystem/FileSystemManager.cs

```diff
@@ -0,0 +1,263 @@
+                        // should failed with IOException
+                        // (Behavior verified on desktop .NET)
+                        //
+                        // Previous FileShare   Requested FileAccess    Requested FileShare
+                        // Read                 Read                    ReadWrite
+                        // Write                Write                   ReadWrite
+                        // ReadWrite            Read                    ReadWrite
+                        // ReadWrite            Write                   ReadWrite
+                        // ReadWrite            ReadWrite               ReadWrite
+                        //
+                        // The following check take advantage of the fact that the value for
+                        // Read, Write, and ReadWrite in FileAccess enum and FileShare enum are
+                        // identical.
+
+                        if ((share != _fileShareReadWrite)
+                            || ((current.Share & access) != access))
+                        {
+                            throw new IOException(
+                                string.Empty,
+                                (int)IOException.IOExceptionErrorCode.UnauthorizedAccess);
+                        }
+                    }
+                }
+
+                _openFiles.Add(record);
+            }
+
+            return record;
+        }
+
+        public static void RemoveFromOpenList(object record)
+        {
+            lock (_openFiles)
+            {
+                _openFiles.Remove(record);
+            }
+        }
+
+        public static object LockDirectory(string directory)
+        {
+            directory = directory.ToUpper();
+
+            lock (_openFiles)
+            {
+                int count = _openFiles.Count;
+
+                for (int i = 0; i < count; i++)
+                {
+                    if (IsInDirectory(
+                        ((FileRecord)_openFiles[i]).FullName,
+                        directory))
+                    {
+                        throw new IOException(
+                            string.Empty,
+                            (int)IOException.IOExceptionErrorCode.UnauthorizedAccess);
+                    }
+                }
+
+                count = _lockedDirs.Count;
+
+                for (int i = 0; i < count; i++)
+                {
+                    if (((string)_lockedDirs[i]) == directory)
+                    {
+                        throw new IOException(
+                            string.Empty,
+                            (int)IOException.IOExceptionErrorCode.UnauthorizedAccess);
+                    }
+                }
+
+                _lockedDirs.Add(directory);
+            }
+
+            return (object)directory;
+        }
+
+        public static void UnlockDirectory(object record)
+        {
+            lock (_openFiles)
```

Comment:
should this be locked on _lockedDirs array?

Reply:
This is to prevent "messing" with a directory when files are open.

Reply:
Might be worth commenting that the lock is due to the Dir being already open, so we also lock the files open?